### ISSUE TITLE
@geometry is not working

### ISSUE THREAD

**Issue by Mario-F-F** [qgis/QGIS#50791]
### What is the bug or the crash?

The "new variable" @geometry is not working when creating a new field

The preview shows the righ value, but, once "acepting" the resulting field is NULL for all values

### Steps to reproduce the issue

Taking any vector layer, create a new field using an expresion using `@geometry`

This expressión does not work:
`min(( bounds_height(@geometry)*1.1)/0.180, (bounds_width( @geometry)*1.1)/0.277)`

      Previus result info is OK, but the field is NULL for all the features

This works fine
`min((bounds_height( $geometry)*1.1)/0.180, (bounds_width( $geometry)*1.1)/0.277)`

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

Versión de QGIS | 3.28.0-Firenze | Revisión del código de QGIS | ed3ad0430f3
-- | -- | -- | --
Versión Qt | 5.15.3
Versión de Python | 3.9.5
Compilado contra GDAL/OGR | 3.5.2 | Ejecutándose contra GDAL/OGR | 3.5.3
Versión de PROJ | 9.1.0
Versión del registro de base de datos EPSG | v10.074 (2022-08-01)
Versión de GEOS | 3.10.3-CAPI-1.16.1
Versión de SQLite | 3.39.4
Versión de PDAL | 2.4.3
Versión del cliente de PostgreSQL | unknown
Versión de SpatiaLite | 5.0.1
Versión de QWT | 6.1.6
Versión de QScintilla2 | 2.13.1
Versión del SO | Windows 10 Version 2009
  |   |   |  
Complementos activos de Python
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.6
processing | 2.12.99
sagaprovider | 2.12.99

</body></html>Versión de QGIS
3.28.0-Firenze
Revisión del código de QGIS
[ed3ad0430f3](https://github.com/qgis/QGIS/commit/ed3ad0430f3)
Versión Qt
5.15.3
Versión de Python
3.9.5
Compilado contra GDAL/OGR
3.5.2
Ejecutándose contra GDAL/OGR
3.5.3
Versión de PROJ
9.1.0
Versión del registro de base de datos EPSG
v10.074 (2022-08-01)
Versión de GEOS
3.10.3-CAPI-1.16.1
Versión de SQLite
3.39.4
Versión de PDAL
2.4.3
Versión del cliente de PostgreSQL
unknown
Versión de SpatiaLite
5.0.1
Versión de QWT
6.1.6
Versión de QScintilla2
2.13.1
Versión del SO
Windows 10 Version 2009

Complementos activos de Python
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by agiudiceandrea** [comment]
@Mario-F-F, I cannot replicate the issue using a Memory layer. Could you please share a sample layer with a minimum number of feature and specify in detail what kind of field and every parameter used in the field calculator.

**Comment by Mario-F-F** [comment]
I had dificulties to reproduce the issue

I have upload this test project
https://github.com/Mario-F-F/data/blob/c4082dadc23edd3b63ad349b99459c669af91df4/test_at_dollar.gpkg


It only have twice the same layer 'barrios'

barrios_c1 has a virtual field using @geometry
barrios_c2 has two virtual fields using @geometry and $geometry

![imagen](https://user-images.githubusercontent.com/10926064/200165656-44f6ab27-3462-4e6f-82b9-a8e5c28a3ab6.png)

In barrios_c1, E_at is defined as
`if( 
    (bounds_width( @geometry)/bounds_height( @geometry)) < (277.00/180.00),
	( bounds_height( @geometry)*1.1)/0.180,
    ( bounds_width( @geometry)*1.1)/0.277
	)`

All values are null

In barrios_c2, E_at is defined the same way, and the values are ok

**Comment by phidrho** [comment]
Hi @agiudiceandrea, @Mario-F-F, @nyalldawson,

I also came to this bug today, it's still present in latest published ver: 3.18.3 on Windows. This happened to me on GPKG and on PostGIS layer, but doesn't affect scratch layers.
I did not use virtual fields, I tried to create regular fields. Inside expression editor everything seems fine, but after executing expression, it results with NULL records. I did not try with new profile, but I have a fresh install of QGIS without any plugin added.

![calculator_gpkg](https://user-images.githubusercontent.com/1849293/215568088-5bdabe94-2067-4ffe-8d18-e32597db708d.png)

![calculator_gpkg_table](https://user-images.githubusercontent.com/1849293/215568415-7849b40c-024a-4c0b-a305-4aacc3474d11.png)

![calculator_gpkg_dollar](https://user-images.githubusercontent.com/1849293/215568052-9bdc96f7-c6a7-46fc-ab76-d9051a017745.png)

![calculator_gpkg_dollar_table](https://user-images.githubusercontent.com/1849293/215568151-61babfa3-872b-416a-8c35-665f0fa9af13.png)
